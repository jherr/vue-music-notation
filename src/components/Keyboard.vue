<template>
  <svg
    :width="width"
    :height="height"
    style="shape-rendering: crispEdges;"
  >
    <path
      :d="path.path"
      :class="{ [path.className]: true, active: path.active }"
      :key="path.note"
      v-for="path in keys"
    />
  </svg>
</template>

<script>
const KEYSHAPES = [
  { type: 0, right: 1 },          // C
  { type: 1 },                    // C#
  { type: 0, right: 1, left: 1 }, // D
  { type: 1 },                    // D#
  { type: 0, left: 1 },           // E
  { type: 0, right: 1 },          // F
  { type: 1 },                    // F#
  { type: 0, right: 1, left: 1 }, // G
  { type: 1 },                    // G#
  { type: 0, right: 1, left: 1 }, // A
  { type: 1 },                    // A#
  { type: 0, left: 1 },           // B
];
const NATURAL_WIDTH = 13;
const NATURAL_HEIGHT = NATURAL_WIDTH * 5.5;
const SHARP_WIDTH = NATURAL_WIDTH / 2;
const SHARP_HEIGHT = NATURAL_HEIGHT / 2;

export default {
  props: {
    length: {
      default: 88,
    },
    start: {
      default: 21,
    },
    notes: {
      default: [],
    },
  },
  data() {
    const isActive = {};
    this.notes.forEach(n => isActive[n] = true);

    const keys = [];

    let position = -NATURAL_WIDTH;
    position += 1;
    let wasSharp = false;
    for (let note = this.start; note < this.start + this.length; note++) {
      const shape = { ...KEYSHAPES[note % 12] };
      if (note === this.start) {
        delete shape.left;
      }
      if (note === (this.start + this.length) -1) {
        delete shape.right;
      }

      let path = '';
      let className = '';
      if (shape.type === 0) {
        position += wasSharp ? (SHARP_WIDTH / 2) : NATURAL_WIDTH;
        wasSharp = false;

        if (shape.right && shape.left) {
          path = `M ${position} ${NATURAL_HEIGHT + 1} v ${-SHARP_HEIGHT} `;
          path += `h ${(SHARP_WIDTH / 2)} v ${-SHARP_HEIGHT} `;
          path += `h ${NATURAL_WIDTH - SHARP_WIDTH} v ${SHARP_HEIGHT} h ${SHARP_WIDTH / 2} `;
          path += `v ${SHARP_HEIGHT} h ${-NATURAL_WIDTH}`;
        } else if (shape.right) {
          path = `M ${position} ${NATURAL_HEIGHT + 1} v ${-NATURAL_HEIGHT} `;
          path += `h ${NATURAL_WIDTH - (SHARP_WIDTH / 2)} v ${SHARP_HEIGHT} `;
          path += `h ${SHARP_WIDTH / 2} v ${SHARP_HEIGHT} h ${-NATURAL_WIDTH}`;
        } else if (shape.left) {
          path = `M ${position} ${NATURAL_HEIGHT + 1} v ${-SHARP_HEIGHT} `;
          path += `h ${SHARP_WIDTH / 2} v ${-SHARP_HEIGHT} `;
          path += `h ${NATURAL_WIDTH - (SHARP_WIDTH / 2)} v ${NATURAL_HEIGHT} h ${-NATURAL_WIDTH}`;
        } else {
          path = `M ${position} ${NATURAL_HEIGHT + 1} v ${-NATURAL_HEIGHT} `;
          path += `h ${NATURAL_WIDTH} v ${NATURAL_HEIGHT} h ${-NATURAL_WIDTH}`;
        }
        className = 'natural';
      } else {
        position += NATURAL_WIDTH - (SHARP_WIDTH / 2);

        path = `M ${position} ${SHARP_HEIGHT + 1} v ${-SHARP_HEIGHT} `;
        path += `h ${SHARP_WIDTH} v ${SHARP_HEIGHT} h ${-SHARP_WIDTH}`;

        wasSharp = true;
        className = 'sharp';
      }

      keys.push({
        path,
        className,
        note,
        active: isActive[note] !== undefined,
      });
    }

    return {
      keys,
      height: NATURAL_HEIGHT + 2,
      width: position + NATURAL_WIDTH + 2,
    };
  }
}
</script>

<style scoped>
.natural {
  fill: white;
  stroke-width: 1;
  stroke: black;
}

.sharp {
  fill: black;
  stroke-width: 1;
  stroke: black;
}

.active {
  fill: lightblue;
}
</style>