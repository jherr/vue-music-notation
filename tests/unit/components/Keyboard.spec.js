import { mount } from '@vue/test-utils';
import Keyboard from '@/components/Keyboard.vue';

describe('Keyboard', () => {
  it('renders', () => {
    const wrapper = mount(Keyboard, {
      propsData: {
        notes: [ 60, 70, 80],
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });
});
