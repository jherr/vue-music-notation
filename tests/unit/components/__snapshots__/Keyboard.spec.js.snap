// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Keyboard renders 1`] = `
<svg
  height="73.5"
  style="shape-rendering: crispEdges;"
  width="679"
>
  <path
    class="natural"
    d="M 1 72.5 v -71.5 h 9.75 v 35.75 h 3.25 v 35.75 h -13"
  />
  <path
    class="sharp"
    d="M 10.75 36.75 v -35.75 h 6.5 v 35.75 h -6.5"
  />
  <path
    class="natural"
    d="M 14 72.5 v -35.75 h 3.25 v -35.75 h 9.75 v 71.5 h -13"
  />
  <path
    class="natural"
    d="M 27 72.5 v -71.5 h 9.75 v 35.75 h 3.25 v 35.75 h -13"
  />
  <path
    class="sharp"
    d="M 36.75 36.75 v -35.75 h 6.5 v 35.75 h -6.5"
  />
  <path
    class="natural"
    d="M 40 72.5 v -35.75 h 3.25 v -35.75 h 6.5 v 35.75 h 3.25 v 35.75 h -13"
  />
  <path
    class="sharp"
    d="M 49.75 36.75 v -35.75 h 6.5 v 35.75 h -6.5"
  />
  <path
    class="natural"
    d="M 53 72.5 v -35.75 h 3.25 v -35.75 h 9.75 v 71.5 h -13"
  />
  <path
    class="natural"
    d="M 66 72.5 v -71.5 h 9.75 v 35.75 h 3.25 v 35.75 h -13"
  />
  <path
    class="sharp"
    d="M 75.75 36.75 v -35.75 h 6.5 v 35.75 h -6.5"
  />
  <path
    class="natural"
    d="M 79 72.5 v -35.75 h 3.25 v -35.75 h 6.5 v 35.75 h 3.25 v 35.75 h -13"
  />
  <path
    class="sharp"
    d="M 88.75 36.75 v -35.75 h 6.5 v 35.75 h -6.5"
  />
  <path
    class="natural"
    d="M 92 72.5 v -35.75 h 3.25 v -35.75 h 6.5 v 35.75 h 3.25 v 35.75 h -13"
  />
  <path
    class="sharp"
    d="M 101.75 36.75 v -35.75 h 6.5 v 35.75 h -6.5"
  />
  <path
    class="natural"
    d="M 105 72.5 v -35.75 h 3.25 v -35.75 h 9.75 v 71.5 h -13"
  />
  <path
    class="natural"
    d="M 118 72.5 v -71.5 h 9.75 v 35.75 h 3.25 v 35.75 h -13"
  />
  <path
    class="sharp"
    d="M 127.75 36.75 v -35.75 h 6.5 v 35.75 h -6.5"
  />
  <path
    class="natural"
    d="M 131 72.5 v -35.75 h 3.25 v -35.75 h 6.5 v 35.75 h 3.25 v 35.75 h -13"
  />
  <path
    class="sharp"
    d="M 140.75 36.75 v -35.75 h 6.5 v 35.75 h -6.5"
  />
  <path
    class="natural"
    d="M 144 72.5 v -35.75 h 3.25 v -35.75 h 9.75 v 71.5 h -13"
  />
  <path
    class="natural"
    d="M 157 72.5 v -71.5 h 9.75 v 35.75 h 3.25 v 35.75 h -13"
  />
  <path
    class="sharp"
    d="M 166.75 36.75 v -35.75 h 6.5 v 35.75 h -6.5"
  />
  <path
    class="natural"
    d="M 170 72.5 v -35.75 h 3.25 v -35.75 h 6.5 v 35.75 h 3.25 v 35.75 h -13"
  />
  <path
    class="sharp"
    d="M 179.75 36.75 v -35.75 h 6.5 v 35.75 h -6.5"
  />
  <path
    class="natural"
    d="M 183 72.5 v -35.75 h 3.25 v -35.75 h 6.5 v 35.75 h 3.25 v 35.75 h -13"
  />
  <path
    class="sharp"
    d="M 192.75 36.75 v -35.75 h 6.5 v 35.75 h -6.5"
  />
  <path
    class="natural"
    d="M 196 72.5 v -35.75 h 3.25 v -35.75 h 9.75 v 71.5 h -13"
  />
  <path
    class="natural"
    d="M 209 72.5 v -71.5 h 9.75 v 35.75 h 3.25 v 35.75 h -13"
  />
  <path
    class="sharp"
    d="M 218.75 36.75 v -35.75 h 6.5 v 35.75 h -6.5"
  />
  <path
    class="natural"
    d="M 222 72.5 v -35.75 h 3.25 v -35.75 h 6.5 v 35.75 h 3.25 v 35.75 h -13"
  />
  <path
    class="sharp"
    d="M 231.75 36.75 v -35.75 h 6.5 v 35.75 h -6.5"
  />
  <path
    class="natural"
    d="M 235 72.5 v -35.75 h 3.25 v -35.75 h 9.75 v 71.5 h -13"
  />
  <path
    class="natural"
    d="M 248 72.5 v -71.5 h 9.75 v 35.75 h 3.25 v 35.75 h -13"
  />
  <path
    class="sharp"
    d="M 257.75 36.75 v -35.75 h 6.5 v 35.75 h -6.5"
  />
  <path
    class="natural"
    d="M 261 72.5 v -35.75 h 3.25 v -35.75 h 6.5 v 35.75 h 3.25 v 35.75 h -13"
  />
  <path
    class="sharp"
    d="M 270.75 36.75 v -35.75 h 6.5 v 35.75 h -6.5"
  />
  <path
    class="natural"
    d="M 274 72.5 v -35.75 h 3.25 v -35.75 h 6.5 v 35.75 h 3.25 v 35.75 h -13"
  />
  <path
    class="sharp"
    d="M 283.75 36.75 v -35.75 h 6.5 v 35.75 h -6.5"
  />
  <path
    class="natural"
    d="M 287 72.5 v -35.75 h 3.25 v -35.75 h 9.75 v 71.5 h -13"
  />
  <path
    class="natural active"
    d="M 300 72.5 v -71.5 h 9.75 v 35.75 h 3.25 v 35.75 h -13"
  />
  <path
    class="sharp"
    d="M 309.75 36.75 v -35.75 h 6.5 v 35.75 h -6.5"
  />
  <path
    class="natural"
    d="M 313 72.5 v -35.75 h 3.25 v -35.75 h 6.5 v 35.75 h 3.25 v 35.75 h -13"
  />
  <path
    class="sharp"
    d="M 322.75 36.75 v -35.75 h 6.5 v 35.75 h -6.5"
  />
  <path
    class="natural"
    d="M 326 72.5 v -35.75 h 3.25 v -35.75 h 9.75 v 71.5 h -13"
  />
  <path
    class="natural"
    d="M 339 72.5 v -71.5 h 9.75 v 35.75 h 3.25 v 35.75 h -13"
  />
  <path
    class="sharp"
    d="M 348.75 36.75 v -35.75 h 6.5 v 35.75 h -6.5"
  />
  <path
    class="natural"
    d="M 352 72.5 v -35.75 h 3.25 v -35.75 h 6.5 v 35.75 h 3.25 v 35.75 h -13"
  />
  <path
    class="sharp"
    d="M 361.75 36.75 v -35.75 h 6.5 v 35.75 h -6.5"
  />
  <path
    class="natural"
    d="M 365 72.5 v -35.75 h 3.25 v -35.75 h 6.5 v 35.75 h 3.25 v 35.75 h -13"
  />
  <path
    class="sharp active"
    d="M 374.75 36.75 v -35.75 h 6.5 v 35.75 h -6.5"
  />
  <path
    class="natural"
    d="M 378 72.5 v -35.75 h 3.25 v -35.75 h 9.75 v 71.5 h -13"
  />
  <path
    class="natural"
    d="M 391 72.5 v -71.5 h 9.75 v 35.75 h 3.25 v 35.75 h -13"
  />
  <path
    class="sharp"
    d="M 400.75 36.75 v -35.75 h 6.5 v 35.75 h -6.5"
  />
  <path
    class="natural"
    d="M 404 72.5 v -35.75 h 3.25 v -35.75 h 6.5 v 35.75 h 3.25 v 35.75 h -13"
  />
  <path
    class="sharp"
    d="M 413.75 36.75 v -35.75 h 6.5 v 35.75 h -6.5"
  />
  <path
    class="natural"
    d="M 417 72.5 v -35.75 h 3.25 v -35.75 h 9.75 v 71.5 h -13"
  />
  <path
    class="natural"
    d="M 430 72.5 v -71.5 h 9.75 v 35.75 h 3.25 v 35.75 h -13"
  />
  <path
    class="sharp"
    d="M 439.75 36.75 v -35.75 h 6.5 v 35.75 h -6.5"
  />
  <path
    class="natural"
    d="M 443 72.5 v -35.75 h 3.25 v -35.75 h 6.5 v 35.75 h 3.25 v 35.75 h -13"
  />
  <path
    class="sharp active"
    d="M 452.75 36.75 v -35.75 h 6.5 v 35.75 h -6.5"
  />
  <path
    class="natural"
    d="M 456 72.5 v -35.75 h 3.25 v -35.75 h 6.5 v 35.75 h 3.25 v 35.75 h -13"
  />
  <path
    class="sharp"
    d="M 465.75 36.75 v -35.75 h 6.5 v 35.75 h -6.5"
  />
  <path
    class="natural"
    d="M 469 72.5 v -35.75 h 3.25 v -35.75 h 9.75 v 71.5 h -13"
  />
  <path
    class="natural"
    d="M 482 72.5 v -71.5 h 9.75 v 35.75 h 3.25 v 35.75 h -13"
  />
  <path
    class="sharp"
    d="M 491.75 36.75 v -35.75 h 6.5 v 35.75 h -6.5"
  />
  <path
    class="natural"
    d="M 495 72.5 v -35.75 h 3.25 v -35.75 h 6.5 v 35.75 h 3.25 v 35.75 h -13"
  />
  <path
    class="sharp"
    d="M 504.75 36.75 v -35.75 h 6.5 v 35.75 h -6.5"
  />
  <path
    class="natural"
    d="M 508 72.5 v -35.75 h 3.25 v -35.75 h 9.75 v 71.5 h -13"
  />
  <path
    class="natural"
    d="M 521 72.5 v -71.5 h 9.75 v 35.75 h 3.25 v 35.75 h -13"
  />
  <path
    class="sharp"
    d="M 530.75 36.75 v -35.75 h 6.5 v 35.75 h -6.5"
  />
  <path
    class="natural"
    d="M 534 72.5 v -35.75 h 3.25 v -35.75 h 6.5 v 35.75 h 3.25 v 35.75 h -13"
  />
  <path
    class="sharp"
    d="M 543.75 36.75 v -35.75 h 6.5 v 35.75 h -6.5"
  />
  <path
    class="natural"
    d="M 547 72.5 v -35.75 h 3.25 v -35.75 h 6.5 v 35.75 h 3.25 v 35.75 h -13"
  />
  <path
    class="sharp"
    d="M 556.75 36.75 v -35.75 h 6.5 v 35.75 h -6.5"
  />
  <path
    class="natural"
    d="M 560 72.5 v -35.75 h 3.25 v -35.75 h 9.75 v 71.5 h -13"
  />
  <path
    class="natural"
    d="M 573 72.5 v -71.5 h 9.75 v 35.75 h 3.25 v 35.75 h -13"
  />
  <path
    class="sharp"
    d="M 582.75 36.75 v -35.75 h 6.5 v 35.75 h -6.5"
  />
  <path
    class="natural"
    d="M 586 72.5 v -35.75 h 3.25 v -35.75 h 6.5 v 35.75 h 3.25 v 35.75 h -13"
  />
  <path
    class="sharp"
    d="M 595.75 36.75 v -35.75 h 6.5 v 35.75 h -6.5"
  />
  <path
    class="natural"
    d="M 599 72.5 v -35.75 h 3.25 v -35.75 h 9.75 v 71.5 h -13"
  />
  <path
    class="natural"
    d="M 612 72.5 v -71.5 h 9.75 v 35.75 h 3.25 v 35.75 h -13"
  />
  <path
    class="sharp"
    d="M 621.75 36.75 v -35.75 h 6.5 v 35.75 h -6.5"
  />
  <path
    class="natural"
    d="M 625 72.5 v -35.75 h 3.25 v -35.75 h 6.5 v 35.75 h 3.25 v 35.75 h -13"
  />
  <path
    class="sharp"
    d="M 634.75 36.75 v -35.75 h 6.5 v 35.75 h -6.5"
  />
  <path
    class="natural"
    d="M 638 72.5 v -35.75 h 3.25 v -35.75 h 6.5 v 35.75 h 3.25 v 35.75 h -13"
  />
  <path
    class="sharp"
    d="M 647.75 36.75 v -35.75 h 6.5 v 35.75 h -6.5"
  />
  <path
    class="natural"
    d="M 651 72.5 v -35.75 h 3.25 v -35.75 h 9.75 v 71.5 h -13"
  />
  <path
    class="natural"
    d="M 664 72.5 v -71.5 h 13 v 71.5 h -13"
  />
</svg>
`;
